Design.md

```diff
@@ -0,0 +1,179 @@
+    rpc Output (TargetJobId) returns (stream OutputEvent);
+}
+
+message JobStartRequest {
+    string path = 1;
+    repeated string args = 2;
+}
+
+message StopSignalSent {}
+
+message TargetJobId {
+    bytes jobid = 1;
+}
+
+message JobStatus {
+    oneof completed {               // Empty if still running
+        int32   status_code = 2;    // Completed normally
+        int32   signal = 3;         // Terminated by a signal
+    }
+}
+
+message OutputEvent {
+    enum Stream {
+        stdout = 0;
+        stderr = 1;
+    }
+    Stream stream = 1;
+    bytes output = 2;
+}
+
+```
+
+### Start
+
+Starts a new job by spawning a process from given executable path and arguments. Returns a unique job id (UUID v4), that is used to specify the target job for other endpoints. If the executable is not found or cannot be executed, immediately returns an error.
+
+No security checks are applied to the program and arguments. However, as the job is placed inside an isolated container, it shouldn't be able to do much damage. It can still consume resources in the limits that cgroup-limits allow, and it can flood it's output with thrash data.
+
+### Stop
+
+Cancels a job by sending `SIGKILL`. This is done asynchronously, and stop can return before the process has terminated. If the client must wait until the job has stopped, it can do so by polling `Status`.
+
+Could be improved by sending `SIGTERM` shortly before `SIGKILL`, but I'm aiming for simplicity here. Also separating soft and hard kills in the API level might be useful in some situations.
+
+### Status
+
+Returns job status, i.e. is it running, and the status code if the job has completed. If the job has been terminated with a signal, that is reported instead.
+
+### Output
+
+Streams output of a job in binary blobs. Each blob is tagged to be either from stdout or stderr. Stream is automatically closed when the process completes and all output has been streamed. All calls to output stream the whole output history from the moment the process was started.
+
+#### Internals
+
+There will be two async-tasks reading the output of a process, one for stdout and one for stderr. For both tasks, a separate buffer of output history is maintained, along with a boolean marking process completion. This state is protected by a [`tokio::sync::RwLock`](https://docs.rs/tokio/1.6.1/tokio/sync/struct.RwLock.html). In addition, there will be a [`tokio::sync::Notify`](https://docs.rs/tokio/1.6.1/tokio/sync/struct.Notify.html). Every time the process writes more output, the listerners are woken up through the `Notify`.
+
+Each call of `Output` spawns async-tasks for stdio and stdout. They read the output buffer until the end. Then it checks if the process is completed (from the field). If yes, then the connection to client is closed to mark process completion. Otherwise, it waits until the output reader task notifies it that new data is available, and then repeats the above process.
+
+## CLI
```

Comment:
nit: how are the client credentials passed?

Reply:
Current idea was to use a hard coded location or a config file. However, if I need to test with multiple clients, it might be wise to change that to a command line parameter.

Reply:
Env variables will be used. Updated the doc.